### PR TITLE
Ruby Capstone - Catalogue of Things 

### PR DESCRIPTION
**Pull request by killy10o10**
In this project, we created a console app that will help you record different types of things you own: books, music albums, movies, and games. The data will be stored in JSON files. 

catalogue-of-things

The project is a group project that four individuals handled. @aeh1707  @WalterOkumu  @killy10o10  and @iambenkis.

A kanban project board was set up to handle team group tasks and individual task assignments and indicates clear milestones for each individual and for the group as well.

The following was performed in the project as we met the project goals and objectives.

Logic
With the provided UML diagram, each class was created in a separate .rb file.
All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes were implemented.
All properties visible in the diagram was set up in the constructor method except for 1-many relationship where custom setters were used.
All class Methods in the UML diagram were created and Implemented.
User Interface
A main.rb file which serves as the console app entry-point was created and implemented.
Your console app, at the Load,
- Presents the user with a list of options to perform.
- Let users choose an option.
- If needed, ask for parameters for the option and also
- Has the option to exit the app.
The options for listing all the items, as listed below was created in the UI
- List all books
- List all music albums
- List of games
- List all genres (e.g 'Comedy', 'Thriller')
- List all labels (e.g. 'Gift', 'New')
- List all authors (e.g. 'Stephen King')
- List all sources (e.g. 'From a friend, 'Online shop')
The option to add items to the list was also created in the UI, these options are,
- Add a book
- Add a music album
- Add a game
- List all movies
- List all sources (e.g. 'From a friend', 'Online shop')
- Add a movie
Data preservation and storage were also performed by using file i/o to store data as .json files.
Unit Testing
This indicated the final milestone of the project as testing using RSpec was performed.

Testing of each class was performed and indicated passed.
Testing of class methods, such as add_item(), can_be_archived?(), was performed as well.

**Kindly find the project presentation [here](https://youtu.be/341TEfD5vU4)** 😄 🚀

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,6 +36,10 @@
 ###UML Class Diagram
 ![uml-class-diagram](uml-class-diagram.png)
 
+## Demo
+
+> **Video Presentation** <iframe width="560" height="315" src="https://www.youtube.com/embed/341TEfD5vU4" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Comment (Ismailco) [inline README.md:41]: - [ ] The video should be 5-minute max, and you should present some piece of code you are proud of making, kindly make sure your video respect these requirements

File: sources/source.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# i2 = Item.new('2007')
+# i3 = Item.new('2020')
+# i4 = Item.new('2022')
+# s = Source.new('Ben')
+# s.add_item(i1)
+# puts s.items
+# s.add_item(i2)
+# s.add_item(i3)
+# s.add_item(i4)
+# puts s.items[3].publish_date
```
Comment (Ismailco) [inline sources/source.rb:29]: - [ ] Kindly consider removing all unused pieces of code from your files they make your code messy and harder to read

File: music_albums/music_album.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      \"on_spotify\": \"#{on_spotify}\",
+      \"publish_date\": \"#{publish_date}\"
+    }"
+  end
+
+  # private
+
+  def can_be_archived?()
+    parent = super()
+    return true if parent || @on_spotify
```
Comment (Ismailco) [inline music_albums/music_album.rb:22]: - [ ] `can_be_archived?` method should return true if the parent's method returns true **AND** if `@on_spotify` equals true, kindly pay attention here your using the `||` **OR** symbol. make sure to adjust your method so it returns the proper value according to the above condition

File: data.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+def save_source(source)
+  JSON.dump(source, File.open('./storage/source.json', 'w+'))
+end
+
+def save_data(item)
+  save_album(item.music_albums)
+  save_genre(item.genres)
+  save_source(item.sources)
+  save_labels(item.labels)
+end
```
Comment (Ismailco) [inline data.rb:24]: - [ ] When I try to create a new game, movie, or music album... after I removed the `storage` folder the app crashes, kindly consider fixing this so the app saves the data even when the folder is messing

Screenshot_20221126_133347

Screenshot_20221126_133504

**Review by Ismailco — CHANGES_REQUESTED** [review]
#### Hi @killy10o10, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!
You did a great job, keep up the good work 👍🏼

giphy


## Highlights

✔️  No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: README.md (hunk 1 above)
Comment (mahbubul14) [reply README.md:41]: > * [x]  The video should be 5-minute max, and you should present some piece of code you are proud of making, kindly make sure your video respect these requirements

- [ ] As the previous reviewer requested, Kindly work on it. :+1:

**Review by mahbubul14 — CHANGES_REQUESTED** [review]
**Hi @Team,**

Good job so far! :+1: 
There are some issues that you still need to work on to go to the next project but you are almost there!

# Required Changes ♻️

- [ ] Kindly check the inline comments under the review. :arrow_down: 

###  To Highlights
:heavy_check_mark: No linter error.
:heavy_check_mark: Proper Github flow used.
:heavy_check_mark: Descriptive PR title & description.

### Optional suggestions

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better.

* N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification. You can also reach me on [slack](https://microverse-students.slack.com/team/U01J39GL614)

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Review by danielufeli — APPROVED** [review]
Hi @Team,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

ChopApproveCopApproveGIF

## Highlights
- Nice commits messages✔️
- Descriptive PR ✔️
- Clean Code✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @danielufeli in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._ate_comments_counter`

**Pull request merged**